// lemmy-js-client/dist/types/others.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserOperation = exports.VERSION = void 0;
exports.VERSION = "v3";
/**
 * All of the websocket operations available.
 */
var UserOperation;
(function (UserOperation) {
    UserOperation[UserOperation["Login"] = 0] = "Login";
    UserOperation[UserOperation["Register"] = 1] = "Register";
    UserOperation[UserOperation["GetCaptcha"] = 2] = "GetCaptcha";
    UserOperation[UserOperation["CreateCommunity"] = 3] = "CreateCommunity";
    UserOperation[UserOperation["CreatePost"] = 4] = "CreatePost";
    UserOperation[UserOperation["ListCommunities"] = 5] = "ListCommunities";
    UserOperation[UserOperation["GetPost"] = 6] = "GetPost";
    UserOperation[UserOperation["GetCommunity"] = 7] = "GetCommunity";
    UserOperation[UserOperation["CreateComment"] = 8] = "CreateComment";
    UserOperation[UserOperation["EditComment"] = 9] = "EditComment";
    UserOperation[UserOperation["DeleteComment"] = 10] = "DeleteComment";
    UserOperation[UserOperation["RemoveComment"] = 11] = "RemoveComment";
    UserOperation[UserOperation["SaveComment"] = 12] = "SaveComment";
    UserOperation[UserOperation["CreateCommentLike"] = 13] = "CreateCommentLike";
    UserOperation[UserOperation["GetPosts"] = 14] = "GetPosts";
    UserOperation[UserOperation["CreatePostLike"] = 15] = "CreatePostLike";
    UserOperation[UserOperation["EditPost"] = 16] = "EditPost";
    UserOperation[UserOperation["DeletePost"] = 17] = "DeletePost";
    UserOperation[UserOperation["RemovePost"] = 18] = "RemovePost";
    UserOperation[UserOperation["LockPost"] = 19] = "LockPost";
    UserOperation[UserOperation["FeaturePost"] = 20] = "FeaturePost";
    UserOperation[UserOperation["MarkPostAsRead"] = 21] = "MarkPostAsRead";
    UserOperation[UserOperation["SavePost"] = 22] = "SavePost";
    UserOperation[UserOperation["EditCommunity"] = 23] = "EditCommunity";
    UserOperation[UserOperation["DeleteCommunity"] = 24] = "DeleteCommunity";
    UserOperation[UserOperation["RemoveCommunity"] = 25] = "RemoveCommunity";
    UserOperation[UserOperation["FollowCommunity"] = 26] = "FollowCommunity";
    UserOperation[UserOperation["GetPersonDetails"] = 27] = "GetPersonDetails";
    UserOperation[UserOperation["GetReplies"] = 28] = "GetReplies";
    UserOperation[UserOperation["GetPersonMentions"] = 29] = "GetPersonMentions";
    UserOperation[UserOperation["MarkPersonMentionAsRead"] = 30] = "MarkPersonMentionAsRead";
    UserOperation[UserOperation["MarkCommentReplyAsRead"] = 31] = "MarkCommentReplyAsRead";
    UserOperation[UserOperation["GetModlog"] = 32] = "GetModlog";
    UserOperation[UserOperation["BanFromCommunity"] = 33] = "BanFromCommunity";
    UserOperation[UserOperation["AddModToCommunity"] = 34] = "AddModToCommunity";
    UserOperation[UserOperation["CreateSite"] = 35] = "CreateSite";
    UserOperation[UserOperation["EditSite"] = 36] = "EditSite";
    UserOperation[UserOperation["GetSite"] = 37] = "GetSite";
    UserOperation[UserOperation["AddAdmin"] = 38] = "AddAdmin";
    UserOperation[UserOperation["GetUnreadRegistrationApplicationCount"] = 39] = "GetUnreadRegistrationApplicationCount";
    UserOperation[UserOperation["ListRegistrationApplications"] = 40] = "ListRegistrationApplications";
    UserOperation[UserOperation["ApproveRegistrationApplication"] = 41] = "ApproveRegistrationApplication";
    UserOperation[UserOperation["BanPerson"] = 42] = "BanPerson";
    UserOperation[UserOperation["GetBannedPersons"] = 43] = "GetBannedPersons";
    UserOperation[UserOperation["Search"] = 44] = "Search";
    UserOperation[UserOperation["ResolveObject"] = 45] = "ResolveObject";
    UserOperation[UserOperation["MarkAllAsRead"] = 46] = "MarkAllAsRead";
    UserOperation[UserOperation["SaveUserSettings"] = 47] = "SaveUserSettings";
    UserOperation[UserOperation["TransferCommunity"] = 48] = "TransferCommunity";
    UserOperation[UserOperation["LeaveAdmin"] = 49] = "LeaveAdmin";
    UserOperation[UserOperation["DeleteAccount"] = 50] = "DeleteAccount";
    UserOperation[UserOperation["PasswordReset"] = 51] = "PasswordReset";
    UserOperation[UserOperation["PasswordChangeAfterReset"] = 52] = "PasswordChangeAfterReset";
    UserOperation[UserOperation["CreatePrivateMessage"] = 53] = "CreatePrivateMessage";
    UserOperation[UserOperation["EditPrivateMessage"] = 54] = "EditPrivateMessage";
    UserOperation[UserOperation["DeletePrivateMessage"] = 55] = "DeletePrivateMessage";
    UserOperation[UserOperation["MarkPrivateMessageAsRead"] = 56] = "MarkPrivateMessageAsRead";
    UserOperation[UserOperation["CreatePrivateMessageReport"] = 57] = "CreatePrivateMessageReport";
    UserOperation[UserOperation["ResolvePrivateMessageReport"] = 58] = "ResolvePrivateMessageReport";
    UserOperation[UserOperation["ListPrivateMessageReports"] = 59] = "ListPrivateMessageReports";
    UserOperation[UserOperation["GetPrivateMessages"] = 60] = "GetPrivateMessages";
    UserOperation[UserOperation["UserJoin"] = 61] = "UserJoin";
    UserOperation[UserOperation["GetComments"] = 62] = "GetComments";
    UserOperation[UserOperation["GetComment"] = 63] = "GetComment";
    UserOperation[UserOperation["PostJoin"] = 64] = "PostJoin";
    UserOperation[UserOperation["CommunityJoin"] = 65] = "CommunityJoin";
    UserOperation[UserOperation["ChangePassword"] = 66] = "ChangePassword";
    UserOperation[UserOperation["GetSiteMetadata"] = 67] = "GetSiteMetadata";
    UserOperation[UserOperation["BlockCommunity"] = 68] = "BlockCommunity";
    UserOperation[UserOperation["BlockPerson"] = 69] = "BlockPerson";
    UserOperation[UserOperation["PurgePerson"] = 70] = "PurgePerson";
    UserOperation[UserOperation["PurgeCommunity"] = 71] = "PurgeCommunity";
    UserOperation[UserOperation["PurgePost"] = 72] = "PurgePost";
    UserOperation[UserOperation["PurgeComment"] = 73] = "PurgeComment";
    UserOperation[UserOperation["CreateCommentReport"] = 74] = "CreateCommentReport";
    UserOperation[UserOperation["ResolveCommentReport"] = 75] = "ResolveCommentReport";
    UserOperation[UserOperation["ListCommentReports"] = 76] = "ListCommentReports";
    UserOperation[UserOperation["CreatePostReport"] = 77] = "CreatePostReport";
    UserOperation[UserOperation["ResolvePostReport"] = 78] = "ResolvePostReport";
    UserOperation[UserOperation["ListPostReports"] = 79] = "ListPostReports";
    UserOperation[UserOperation["GetReportCount"] = 80] = "GetReportCount";
    UserOperation[UserOperation["GetUnreadCount"] = 81] = "GetUnreadCount";
    UserOperation[UserOperation["VerifyEmail"] = 82] = "VerifyEmail";
    UserOperation[UserOperation["CreateCustomEmoji"] = 83] = "CreateCustomEmoji";
    UserOperation[UserOperation["EditCustomEmoji"] = 84] = "EditCustomEmoji";
    UserOperation[UserOperation["DeleteCustomEmoji"] = 85] = "DeleteCustomEmoji";
    UserOperation[UserOperation["GetFederatedInstances"] = 86] = "GetFederatedInstances";
})(UserOperation = exports.UserOperation || (exports.UserOperation = {}));
