// lemmy-js-client/dist/websocket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wsJsonToRes = exports.wsUserOp = exports.LemmyWebsocket = void 0;
var others_1 = require("./types/others");
// @ts-ignore
BigInt.prototype["toJSON"] = function () {
    return parseInt(this.toString());
};
/**
 * Helps build lemmy websocket message requests, that you can use in your Websocket sends.
 *
 * You'll receive back a [[WebSocketResponse]].
 *
 * The return types for these are given in [[LemmyHttp]]
 */
var LemmyWebsocket = /** @class */ (function () {
    function LemmyWebsocket() {
    }
    /**
     * Log into lemmy.
     */
    LemmyWebsocket.prototype.login = function (form) {
        return wrapper(others_1.UserOperation.Login, form);
    };
    /**
     * A websocket join for your user.
     *
     * Allows your user to receive private messages and notifications.
     */
    LemmyWebsocket.prototype.userJoin = function (form) {
        return wrapper(others_1.UserOperation.UserJoin, form);
    };
    /**
     * A websocket join for the current post room.
     *
     * Allows your user to receive new comments and updates for that post.
     */
    LemmyWebsocket.prototype.postJoin = function (form) {
        return wrapper(others_1.UserOperation.PostJoin, form);
    };
    /**
     * A websocket join for a given community.
     *
     * Allows your user to receive community updates.
     *
     * Note: community_id: 0, is your front page.
     */
    LemmyWebsocket.prototype.communityJoin = function (form) {
        return wrapper(others_1.UserOperation.CommunityJoin, form);
    };
    /**
     * Register a new user.
     */
    LemmyWebsocket.prototype.register = function (register) {
        return wrapper(others_1.UserOperation.Register, register);
    };
    /**
     * Fetch a Captcha.
     */
    LemmyWebsocket.prototype.getCaptcha = function (form) {
        return wrapper(others_1.UserOperation.GetCaptcha, form);
    };
    /**
     * Create a new community.
     */
    LemmyWebsocket.prototype.createCommunity = function (form) {
        return wrapper(others_1.UserOperation.CreateCommunity, form);
    };
    /**
     * Edit a community.
     */
    LemmyWebsocket.prototype.editCommunity = function (form) {
        return wrapper(others_1.UserOperation.EditCommunity, form);
    };
    /**
     * Delete a community.
     */
    LemmyWebsocket.prototype.deleteCommunity = function (form) {
        return wrapper(others_1.UserOperation.DeleteCommunity, form);
    };
    /**
     * A moderator remove for a community.
     */
    LemmyWebsocket.prototype.removeCommunity = function (form) {
        return wrapper(others_1.UserOperation.RemoveCommunity, form);
    };
    /**
     * Follow / subscribe to a community.
     */
    LemmyWebsocket.prototype.followCommunity = function (form) {
        return wrapper(others_1.UserOperation.FollowCommunity, form);
    };
    /**
     * List communities, with various filters.
     */
    LemmyWebsocket.prototype.listCommunities = function (form) {
        return wrapper(others_1.UserOperation.ListCommunities, form);
    };
    /**
     * Create a post.
     */
    LemmyWebsocket.prototype.createPost = function (form) {
        return wrapper(others_1.UserOperation.CreatePost, form);
    };
    /**
     * Get / fetch a post.
     */
    LemmyWebsocket.prototype.getPost = function (form) {
        return wrapper(others_1.UserOperation.GetPost, form);
    };
    /**
     * Get / fetch a community.
     */
    LemmyWebsocket.prototype.getCommunity = function (form) {
        return wrapper(others_1.UserOperation.GetCommunity, form);
    };
    /**
     * Create a comment.
     */
    LemmyWebsocket.prototype.createComment = function (form) {
        return wrapper(others_1.UserOperation.CreateComment, form);
    };
    /**
     * Edit a comment.
     */
    LemmyWebsocket.prototype.editComment = function (form) {
        return wrapper(others_1.UserOperation.EditComment, form);
    };
    /**
     * Delete a comment.
     */
    LemmyWebsocket.prototype.deleteComment = function (form) {
        return wrapper(others_1.UserOperation.DeleteComment, form);
    };
    /**
     * A moderator remove for a comment.
     */
    LemmyWebsocket.prototype.removeComment = function (form) {
        return wrapper(others_1.UserOperation.RemoveComment, form);
    };
    /**
     * Mark a comment as read.
     */
    LemmyWebsocket.prototype.markCommentReplyAsRead = function (form) {
        return wrapper(others_1.UserOperation.MarkCommentReplyAsRead, form);
    };
    /**
     * Like / vote on a comment.
     */
    LemmyWebsocket.prototype.likeComment = function (form) {
        return wrapper(others_1.UserOperation.CreateCommentLike, form);
    };
    /**
     * Save a comment.
     */
    LemmyWebsocket.prototype.saveComment = function (form) {
        return wrapper(others_1.UserOperation.SaveComment, form);
    };
    /**
     * Report a comment.
     */
    LemmyWebsocket.prototype.createCommentReport = function (form) {
        return wrapper(others_1.UserOperation.CreateCommentReport, form);
    };
    /**
     * Resolve a comment report. Only a mod can do this.
     */
    LemmyWebsocket.prototype.resolveCommentReport = function (form) {
        return wrapper(others_1.UserOperation.ResolveCommentReport, form);
    };
    /**
     * List comment reports.
     */
    LemmyWebsocket.prototype.listCommentReports = function (form) {
        return wrapper(others_1.UserOperation.ListCommentReports, form);
    };
    /**
     * Get / fetch posts, with various filters.
     */
    LemmyWebsocket.prototype.getPosts = function (form) {
        return wrapper(others_1.UserOperation.GetPosts, form);
    };
    /**
     * Get / fetch comments.
     */
    LemmyWebsocket.prototype.getComments = function (form) {
        return wrapper(others_1.UserOperation.GetComments, form);
    };
    /**
     * Get / fetch comment.
     */
    LemmyWebsocket.prototype.getComment = function (form) {
        return wrapper(others_1.UserOperation.GetComment, form);
    };
    /**
     * Like / vote on a post.
     */
    LemmyWebsocket.prototype.likePost = function (form) {
        return wrapper(others_1.UserOperation.CreatePostLike, form);
    };
    /**
     * Edit a post.
     */
    LemmyWebsocket.prototype.editPost = function (form) {
        return wrapper(others_1.UserOperation.EditPost, form);
    };
    /**
     * Delete a post.
     */
    LemmyWebsocket.prototype.deletePost = function (form) {
        return wrapper(others_1.UserOperation.DeletePost, form);
    };
    /**
     * A moderator remove for a post.
     */
    LemmyWebsocket.prototype.removePost = function (form) {
        return wrapper(others_1.UserOperation.RemovePost, form);
    };
    /**
     * A moderator can lock a post ( IE disable new comments ).
     */
    LemmyWebsocket.prototype.lockPost = function (form) {
        return wrapper(others_1.UserOperation.LockPost, form);
    };
    /**
     * A moderator can feature a post ( IE stick it to the top of a community ).
     */
    LemmyWebsocket.prototype.featurePost = function (form) {
        return wrapper(others_1.UserOperation.FeaturePost, form);
    };
    /**
     * Mark a post as read.
     */
    LemmyWebsocket.prototype.markPostAsRead = function (form) {
        return wrapper(others_1.UserOperation.MarkPostAsRead, form);
    };
    /**
     * Save a post.
     */
    LemmyWebsocket.prototype.savePost = function (form) {
        return wrapper(others_1.UserOperation.SavePost, form);
    };
    /**
     * Report a post.
     */
    LemmyWebsocket.prototype.createPostReport = function (form) {
        return wrapper(others_1.UserOperation.CreatePostReport, form);
    };
    /**
     * Resolve a post report. Only a mod can do this.
     */
    LemmyWebsocket.prototype.resolvePostReport = function (form) {
        return wrapper(others_1.UserOperation.ResolvePostReport, form);
    };
    /**
     * List post reports.
     */
    LemmyWebsocket.prototype.listPostReports = function (form) {
        return wrapper(others_1.UserOperation.ListPostReports, form);
    };
    /**
     * Fetch metadata for any given site.
     */
    LemmyWebsocket.prototype.getSiteMetadata = function (form) {
        return wrapper(others_1.UserOperation.GetSiteMetadata, form);
    };
    /**
     * Ban a user from a community.
     */
    LemmyWebsocket.prototype.banFromCommunity = function (form) {
        return wrapper(others_1.UserOperation.BanFromCommunity, form);
    };
    /**
     * Add a moderator to your community.
     */
    LemmyWebsocket.prototype.addModToCommunity = function (form) {
        return wrapper(others_1.UserOperation.AddModToCommunity, form);
    };
    /**
     * Transfer your community to an existing moderator.
     */
    LemmyWebsocket.prototype.transferCommunity = function (form) {
        return wrapper(others_1.UserOperation.TransferCommunity, form);
    };
    /**
     * Leave the Site admins.
     */
    LemmyWebsocket.prototype.leaveAdmin = function (form) {
        return wrapper(others_1.UserOperation.LeaveAdmin, form);
    };
    /**
     * Ban a person from your site.
     */
    LemmyWebsocket.prototype.banPerson = function (form) {
        return wrapper(others_1.UserOperation.BanPerson, form);
    };
    /**
     * Get a list of banned users
     */
    LemmyWebsocket.prototype.getBannedPersons = function (form) {
        return wrapper(others_1.UserOperation.GetBannedPersons, form);
    };
    /**
     * Add an admin to your site.
     */
    LemmyWebsocket.prototype.addAdmin = function (form) {
        return wrapper(others_1.UserOperation.AddAdmin, form);
    };
    /**
     * Get the unread registration applications count.
     */
    LemmyWebsocket.prototype.getUnreadRegistrationApplicationCount = function (form) {
        return wrapper(others_1.UserOperation.GetUnreadRegistrationApplicationCount, form);
    };
    /**
     * List the unread registration applications.
     */
    LemmyWebsocket.prototype.listRegistrationApplications = function (form) {
        return wrapper(others_1.UserOperation.ListRegistrationApplications, form);
    };
    /**
     * Approve a registration application
     */
    LemmyWebsocket.prototype.approveRegistrationApplication = function (form) {
        return wrapper(others_1.UserOperation.ApproveRegistrationApplication, form);
    };
    /**
     * Get the details for a person.
     */
    LemmyWebsocket.prototype.getPersonDetails = function (form) {
        return wrapper(others_1.UserOperation.GetPersonDetails, form);
    };
    /**
     * Get comment replies.
     */
    LemmyWebsocket.prototype.getReplies = function (form) {
        return wrapper(others_1.UserOperation.GetReplies, form);
    };
    /**
     * Get mentions for your user.
     */
    LemmyWebsocket.prototype.getPersonMentions = function (form) {
        return wrapper(others_1.UserOperation.GetPersonMentions, form);
    };
    /**
     * Mark a person mention as read.
     */
    LemmyWebsocket.prototype.markPersonMentionAsRead = function (form) {
        return wrapper(others_1.UserOperation.MarkPersonMentionAsRead, form);
    };
    /**
     * Get the modlog.
     */
    LemmyWebsocket.prototype.getModlog = function (form) {
        return wrapper(others_1.UserOperation.GetModlog, form);
    };
    /**
     * Create your site.
     */
    LemmyWebsocket.prototype.createSite = function (form) {
        return wrapper(others_1.UserOperation.CreateSite, form);
    };
    /**
     * Edit your site.
     */
    LemmyWebsocket.prototype.editSite = function (form) {
        return wrapper(others_1.UserOperation.EditSite, form);
    };
    /**
     * Gets the site, and your user data.
     */
    LemmyWebsocket.prototype.getSite = function (form) {
        return wrapper(others_1.UserOperation.GetSite, form);
    };
    /**
     * Search lemmy.
     */
    LemmyWebsocket.prototype.search = function (form) {
        return wrapper(others_1.UserOperation.Search, form);
    };
    /**
     * Fetch a non-local / federated object.
     */
    LemmyWebsocket.prototype.resolveObject = function (form) {
        return wrapper(others_1.UserOperation.ResolveObject, form);
    };
    /**
     * Mark all replies as read.
     */
    LemmyWebsocket.prototype.markAllAsRead = function (form) {
        return wrapper(others_1.UserOperation.MarkAllAsRead, form);
    };
    /**
     * Save your user settings.
     */
    LemmyWebsocket.prototype.saveUserSettings = function (form) {
        return wrapper(others_1.UserOperation.SaveUserSettings, form);
    };
    /**
     * Change your user password.
     */
    LemmyWebsocket.prototype.changePassword = function (form) {
        return wrapper(others_1.UserOperation.ChangePassword, form);
    };
    /**
     * Get counts for your reports
     */
    LemmyWebsocket.prototype.getReportCount = function (form) {
        return wrapper(others_1.UserOperation.GetReportCount, form);
    };
    /**
     * Get your unread counts
     */
    LemmyWebsocket.prototype.getUnreadCount = function (form) {
        return wrapper(others_1.UserOperation.GetUnreadCount, form);
    };
    /**
     * Verify your email
     */
    LemmyWebsocket.prototype.verifyEmail = function (form) {
        return wrapper(others_1.UserOperation.VerifyEmail, form);
    };
    /**
     * Delete your account.
     */
    LemmyWebsocket.prototype.deleteAccount = function (form) {
        return wrapper(others_1.UserOperation.DeleteAccount, form);
    };
    /**
     * Reset your password.
     */
    LemmyWebsocket.prototype.passwordReset = function (form) {
        return wrapper(others_1.UserOperation.PasswordReset, form);
    };
    /**
     * Change your password from an email / token based reset.
     */
    LemmyWebsocket.prototype.passwordChange = function (form) {
        return wrapper(others_1.UserOperation.PasswordChangeAfterReset, form);
    };
    /**
     * Create a private message.
     */
    LemmyWebsocket.prototype.createPrivateMessage = function (form) {
        return wrapper(others_1.UserOperation.CreatePrivateMessage, form);
    };
    /**
     * Edit a private message.
     */
    LemmyWebsocket.prototype.editPrivateMessage = function (form) {
        return wrapper(others_1.UserOperation.EditPrivateMessage, form);
    };
    /**
     * Delete a private message.
     */
    LemmyWebsocket.prototype.deletePrivateMessage = function (form) {
        return wrapper(others_1.UserOperation.DeletePrivateMessage, form);
    };
    /**
     * Mark a private message as read.
     */
    LemmyWebsocket.prototype.markPrivateMessageAsRead = function (form) {
        return wrapper(others_1.UserOperation.MarkPrivateMessageAsRead, form);
    };
    /**
     * Get / fetch private messages.
     */
    LemmyWebsocket.prototype.getPrivateMessages = function (form) {
        return wrapper(others_1.UserOperation.GetPrivateMessages, form);
    };
    /**
     * Create a report for a private message.
     */
    LemmyWebsocket.prototype.createPrivateMessageReport = function (form) {
        return wrapper(others_1.UserOperation.CreatePrivateMessageReport, form);
    };
    /**
     * Resolve a report for a private message.
     */
    LemmyWebsocket.prototype.resolvePrivateMessageReport = function (form) {
        return wrapper(others_1.UserOperation.ResolvePrivateMessageReport, form);
    };
    /**
     * List private message reports.
     */
    LemmyWebsocket.prototype.listPrivateMessageReports = function (form) {
        return wrapper(others_1.UserOperation.ListPrivateMessageReports, form);
    };
    /**
     * Block a person.
     */
    LemmyWebsocket.prototype.blockPerson = function (form) {
        return wrapper(others_1.UserOperation.BlockPerson, form);
    };
    /**
     * Block a community.
     */
    LemmyWebsocket.prototype.blockCommunity = function (form) {
        return wrapper(others_1.UserOperation.BlockCommunity, form);
    };
    /**
     * Purge / Delete a person from the database.
     */
    LemmyWebsocket.prototype.purgePerson = function (form) {
        return wrapper(others_1.UserOperation.PurgePerson, form);
    };
    /**
     * Purge / Delete a community from the database.
     */
    LemmyWebsocket.prototype.purgeCommunity = function (form) {
        return wrapper(others_1.UserOperation.PurgeCommunity, form);
    };
    /**
     * Purge / Delete a post from the database.
     */
    LemmyWebsocket.prototype.purgePost = function (form) {
        return wrapper(others_1.UserOperation.PurgePost, form);
    };
    /**
     * Purge / Delete a comment from the database.
     */
    LemmyWebsocket.prototype.purgeComment = function (form) {
        return wrapper(others_1.UserOperation.PurgeComment, form);
    };
    /**
     * Create a custom emoji
     */
    LemmyWebsocket.prototype.createCustomEmoji = function (form) {
        return wrapper(others_1.UserOperation.CreateCustomEmoji, form);
    };
    /**
     * Edit a custom emoji
     */
    LemmyWebsocket.prototype.editCustomEmoji = function (form) {
        return wrapper(others_1.UserOperation.EditCustomEmoji, form);
    };
    /**
     * Delete a custom emoji
     */
    LemmyWebsocket.prototype.deleteCustomEmoji = function (form) {
        return wrapper(others_1.UserOperation.DeleteCustomEmoji, form);
    };
    /**
     * Fetch federated instances.
     */
    LemmyWebsocket.prototype.getFederatedInstances = function (form) {
        return wrapper(others_1.UserOperation.GetFederatedInstances, form);
    };
    return LemmyWebsocket;
}());
exports.LemmyWebsocket = LemmyWebsocket;
function wrapper(op, data) {
    var send = JSON.stringify({ op: others_1.UserOperation[op], data: data });
    return send;
}
function wsUserOp(msg) {
    var opStr = msg.op;
    return others_1.UserOperation[opStr];
}
exports.wsUserOp = wsUserOp;
/**
 * Converts a websocket string response to a usable result
 */
// TODO is this still necessary?
function wsJsonToRes(msg) {
    // Have to deserialize the response again into the correct class
    return msg.data;
}
exports.wsJsonToRes = wsJsonToRes;
